_posts/phoenix-reverse-proxy/2022-02-17-phoenix-reverse-proxy.md

```diff
@@ -0,0 +1,425 @@
+      Module.register_attribute(__MODULE__, :default_reverse_proxy_endpoint, accumulate: false)
+
+      @impl Plug
+      def init(opts) do
+        opts
+      end
+
+      @impl Plug
+      def call(conn, opts) do
+        matching_endpoint = match_endpoint(conn.host, conn.path_info)
+        matching_endpoint.call(conn, matching_endpoint.init(opts))
+      end
+
+      import unquote(PhoenixReverseProxy)
+      use Phoenix.Endpoint, unquote(opts)
+    end
+  end
+  ```
+
+Do not fear! Everything will be explained! 
+
+## Accumulating Routes Phoenix Style
+First off, what are module attributes? Well, most Elixir developers are already using them:
+```elixir
+defmodule M do
+	@x 1
+end
+```
+This sets the attribute `:x` to value `1`. This is syntactic sugar for:
+```elixir
+defmodule M do
+	Module.register_attribute(__MODULE__, :x, accumulate: false)
+	Module.put_attribute(__MODULE__, :x, 1)
+end
+```
+
+There is an option called `accumulate` which means that an attribute can be set multiple times and it will accumulate in a list. Insertion at the head of a list being more efficient for performance means the list will be in the reverse order of the order in which the attributes were specified. Here is an example in `iex`:
+```elixir
+iex(1)> defmodule M do
+...(1)>   Module.register_attribute(__MODULE__, :x, accumulate: true)
+...(1)>   @x 1
+...(1)>   @x 2
+...(1)>   @x 3
+...(1)>
+...(1)>   def get_x, do: @x
+...(1)> end
+{:module, M,
+ <<70, 79, 82, ...>>, {:get_x, 0}}
+iex(2)> M.get_x
+[3, 2, 1] # In the reverse order
+```
+
+If the attributes are desired in the order specified in the module, the list needs to be reversed. The attribute accumulator is used with macros to create a list of routing rules. Only one of the simplest macros is included here to make it easy to understand:
+```elixir
+  defmacro proxy(hostname, endpoint) do
+    quote do
+      @reverse_proxy_routes {
+        unquote(endpoint),
+        unquote(hostname)
+      }
+    end
+  end
+```
+When this macro is called in our `PhoenixReverseProxy` which is a [`Phoenix.Endpoint`](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html) module, it will add an item to the `PhoenixReverseProxy` routing table.
+
+## Harnessing BEAM Pattern Matching Performance
+
+To harness the BEAM pattern matching and get high performance, a generated function that efficiently uses pattern matching is utilised to route the correct [`Phoenix.Endpoint`](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html).
+
+For simplicity, here is a simplified example where the routes are statically defined and where matching functions are generated to match a domain (the real implementation uses accumulated attributes):
+```elixir
+defmodule DomainMatchingExample do
+  @moduledoc """
+  Proxy mappings from hostname to Phoenix.Endpoint
+  """
+  @default_route Default.Endpoint
+  @routes [
+	  {"a.com", A.Endpoint},
+	  {"b.com", B.Endpoint},
+	  {"c.com", C.Endpoint},
+	  {"d.com", D.Endpoint},
```

Comment:
nit: probably d, c, b, a makes more sense here to match the attributes behavior?

Reply:
Good point.